dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java

```diff
@@ -274,29 +275,33 @@ public Label getEndOfThreadLabel(int tid) {
     }
 
     // ----------------------------------------------------------------------------------------------------------------
-    // PTX
+    // GPU
+    public void setOrCreateScopedThread(Arch arch, String name, int id, int ...scopeIds) {
+        ScopeHierarchy scopeHierarchy = switch (arch) {
+            case PTX -> ScopeHierarchy.ScopeHierarchyForPTX(scopeIds[0], scopeIds[1]);
+            case VULKAN -> ScopeHierarchy.ScopeHierarchyForVulkan(scopeIds[0], scopeIds[1], scopeIds[2]);
+            case OPENCL -> ScopeHierarchy.ScopeHierarchyForOpenCL(scopeIds[0], scopeIds[1]);
+            default -> throw new UnsupportedOperationException("Unsupported architecture: " + arch);
+        };
 
-    public void newScopedThread(Arch arch, String name, int id, int ...scopeIds) {
         if(id2FunctionsMap.containsKey(id)) {
-            throw new MalformedProgramException("Function or thread with id " + id + " already exists.");
+            Thread thread = (Thread) id2FunctionsMap.get(id);
+            thread.setScopeHierarchy(scopeHierarchy);
```

Comment:
Are there cases where we really want to set the scope on an already existing thread?

Reply:
This is for the case e.g. we have declaration: 
`{0:r1 = 0;}`
which will create the thread P0 while initializing r1. It happens before we visiting threads and getting to know their scopeHierarchy.

Reply:
The situation should be similar for PTX or Vulkan visitor. How are we handling similar situations there? Are threads visited before initialization?

Reply:
I think we also did it wrong for them. Their declaration will throw a 'thread not exist' exception during register initialization. We did not have a relevant test for this scenario for either PTX or Vulkan visitor.

Reply:
If what you say is true, I would expect an exception in code like [this](https://github.com/hernanponcedeleon/Dat3M/blob/2d6ef05b23f6921d8effe9cee84749071a600cf8/litmus/PTX/Nvidia/Atom-SB.litmus#L5)

Reply:
Ahh I was wrong, in PTX and Vulkan visitor, we visit the thread declarators before visiting the variable declarator.

Reply:
This looks much better now